Adds the reporting Opt In banner

maxlang opened the pull request:
The helpus banner no longer triggers a modal. Clicking "Opt In"
directs you to the "Help Us" page. When on the "Help Us" page
the banner is dismissed for the session. Interacting with the
banner in any way dismisses the banner permanently.

Other notes:
- changes the uiData reducer to create entries for missing keys
- slightly changes the banner styles (tested in old UI)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7180)

<!-- Reviewable:end -->

mrtracy commented on ui/next/app/redux/uiData.ts:
```diff
@@ -153,16 +153,22 @@ export function loadUIData(...keys: string[]) {
-      response.getKeyValues().forEach((val, key) => {
+      let keyValues = response.getKeyValues();
+
+      _.each(keys, (key) => {
```
Does this compile?  getKeyValues() returns a ProtobufMap, I don't think it can be iterated using underscore.

maxlang replied:
`keys` is actually the original list of keys to get, passed in on line 151. I switched to iterating through those keys instead so I could distinguish between "fetched but missing" and "not yet fetched" depending on if the key is present.

mrtracy commented on ui/next/app/redux/uiData.ts:
```diff
@@ -153,16 +153,22 @@ export function loadUIData(...keys: string[]) {
+      let keyValues = response.getKeyValues();
+
+      _.each(keys, (key) => {
+        let decoded: any = null;
```
If you want to have an explicit `setUIDataKey(key, undefined)` to make the not-exists case more explicit, then you don't need to define `decoded` anymore.  Example:

`` typescript
let bb = // get key value if it exists
if (bb) {
  let str = bb.readString(bb.limit);
  if (str) {
    dispatch(setUIDataKey(key, JSON.parse(str));
    return;
  }
  dispatch(setUIDataKey(key, undefined));
}

maxlang replied:
Done.

mrtracy commented on ui/next/app/containers/banner/helpusBanner.tsx:
```diff
@@ -0,0 +1,86 @@
+  context: HelpusBannerContext;
+
+  componentWillMount() {
+    if (this.context && this.context.router && this.context.router.isActive("/help-us/reporting")) {
```
I would move this to the non-persistent UI Data reducer (the one that stores column sorts), and then add an "onEnter" method to the helpus route.

mrtracy replied:
By "this", I mean the dismissed state of this banner.

maxlang replied:
I guess I still don't fully understand the rule of thumb here. I understood why we stored the table sort order in the local ui data reducer object, because as the table is destroyed/recreated we want to preserve the sort.

But this feels so much more self-contained to me. There's only one helpus banner and it's active the entire time. Why should the routing system know about the helpus banner and dismissing it?

mrtracy replied:
And in this case, I would call it something like a "soft dismissal", since there is a persisted "hard" dismissal on the optinAttributes.

mrtracy replied:
Conversely, why should the helpus banner know about the routing system?  To me, this looks way more like spaghetti to me; it's doing conditional logic on a "global" setting which is difficult to access.

maxlang replied:
Changed. I'll admit I'm not thrilled with either solution. Perhaps something better will present itself eventually.

mrtracy commented on ui/next/app/containers/banner/helpusBanner.tsx:
```diff
@@ -0,0 +1,86 @@
+}
+
+let optinAttributes = (state: any): OptInAttributes => state && state.uiData && state.uiData.data && state.uiData.data[KEY_HELPUS];
+let attributesLoaded = (state: any): boolean => state && state.uiData && state.uiData.data && _.has(state.uiData.data, KEY_HELPUS);
```
I don't think this needs to be a separate selector.  Or, it could be derived from optinAttributes:

```typescript
let attributesLoaded = createSelector(
  optinAttributes,
  (attributes) => _.isObject(attributes)
);
```

maxlang replied:
It can't exactly be defined from `optinAttributes` because if `optinAttributes` is undefined, you still don't know if the key is present or not.

mrtracy replied:
Actually, does this even work in combination with your earlier change to the uiData reducer?  i'm not sure `_.has()` will return true of the key's value is undefined instead of null.

mrtracy replied:
Okay, apparently `_.has` does work in that case.  So, my suggestion will not work; the intent is to see if an attempt to load optinAttributes has occurred.

So, I would create a selector just to select uiData.data and reuse it.

```typescript
let uiDataState = (state: any): UIDataSet => state && state.uiData;
let optinAttributes = createSelector(
  uiDataState,
  (state) => state && state.data[KEY_HELPUS])
);
let attributesLoaded = createSelector(
  uiDataState,
  (state) => state && _.has(state.data, KEY_HELPUS)
);
```
It's a little more code, but it removes several `&&` statements.  It also leverages the fact that, if state.uiData exists, it's `data` property is never null.

maxlang replied:
Done.

mrtracy commented on ui/next/app/containers/banner/helpusBanner.tsx:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  render() {
+    let visible: boolean = this.props.attributesLoaded && !this.state.dismissed && (!this.props.attributes || (!this.props.attributes.dismissed && !this.props.attributes.optin));
```
Lots of logic, maybe a comment describing the situation where we display the banner?

maxlang replied:
Done.

tamird commented:
Reviewed 10 of 10 files at r1, 4 of 8 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ui/next/app/redux/uiData.ts, line 161 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDU44lZZyqauJcFh7n:-KKDU44lZZyqauJcFh7o:-1871988983) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/next/app/redux/uiData.ts#L161)):*
> ```XML
>         // Responses from the server return values as ByteBuffer objects, which
>         // represent stringified JSON objects.
>         let bb = keyValues.get(key) && keyValues.get(key).getValue();
> ```

two calls to `get` feels sucky.

---

*[ui/next/app/redux/uiData.ts, line 166 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDUG1HkbA4_qZHVDun:-KKDUG1HkbA4_qZHVDuo:-640313022) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/next/app/redux/uiData.ts#L166)):*
> ```XML
>           if (str) {
>             dispatch(setUIDataKey(key, JSON.parse(str)));
>             return;
> ```

nit this `return` is slightly confusing; i think an `else` will read better since this is a loop-like function. i thought the return would terminate iteration on first reading =/

---

*[ui/styl/partials/layout.styl, line 349 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDUhqo_cC1xv6dlETN:-KKDUhqo_cC1xv6dlETO:1012148505) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/styl/partials/layout.styl#L349)):*
> ```Stylus
>   // only show the first .banner
>   // any .banner with a previous .banner as a sibling is hidden
>   // & ~ &
> ```

commented out code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2, 4 of 8 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180)*
<!-- Sent from Reviewable.io -->

mrtracy commented on ui/next/app/app.tsx:
```diff
@@ -139,7 +140,7 @@ ReactDOM.render(
             <Route path=":table_name" component={ TableDetails } />
           </Route>
         </Route>
-        <Route path="help-us/reporting" component={ HelpUs } />
```
After thinking about this, you could put this in the `componentWillMount` or `componentDidMount` of the HelpUs component.

Commit:
Adds the reporting Opt In banner

The helpus banner no longer triggers a modal. Clicking "Opt In"
directs you to the "Help Us" page. When on the "Help Us" page
the banner is dismissed for the session. Interacting with the
banner in any way dismisses the banner permanently.

Other notes:
- changes the uiData reducer to create entries for missing keys
- slightly changes the banner styles (tested in old UI)

maxlang commented:
@mrtracy moved it to `helpus.tsx`

---

Review status: 3 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[ui/next/app/redux/uiData.ts, line 161 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDU44lZZyqauJcFh7n:-KKKGahZ_vX7UgmOJOdo:1219687417) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/next/app/redux/uiData.ts#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> two calls to `get` feels sucky.
</details>
Changed to `.has` and `.get`

---

*[ui/next/app/redux/uiData.ts, line 166 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDUG1HkbA4_qZHVDun:-KKKGeOxhmxTylBMg-7c:241485963) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/next/app/redux/uiData.ts#L166)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit this `return` is slightly confusing; i think an `else` will read better since this is a loop-like function. i thought the return would terminate iteration on first reading =/
</details>
rewrote it to avoid the nested conditional so the `else` fit in more naturally

---

*[ui/styl/partials/layout.styl, line 349 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-KKDUhqo_cC1xv6dlETN:-KKKGlf-20hHkk-P2K55:-171028065) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd34a56d7893d3a34e66bff17c189fc40926144/ui/styl/partials/layout.styl#L349)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> commented out code?
</details>
Removed. Seems to work fine in the old UI.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180#-:-KKKGpgG_7Jppg7qceFL:1052694266)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7180)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
LGTM
